lib/pos_snapshot_cache.go

```diff
@@ -0,0 +1,240 @@
+package lib
+
+import (
+	"bytes"
+	"github.com/deso-protocol/core/collections"
+	"github.com/dgraph-io/badger/v3"
+	"github.com/pkg/errors"
+	"math"
+	"sort"
+	"sync"
+)
+
+// SnapshotCache is a struct that holds cached versions of the end-of-epoch
+// snapshot data. This data is used to speed up the process of fetching the
+// snapshot data from the database when processing blocks. We only
+// ever update the cache with data from the database and never the view.
+type SnapshotCache struct {
+	// snapshot validator cache
+	cachedValidatorsBySnapshotAtEpoch map[uint64][]*ValidatorEntry
+	validatorRWLock                   sync.RWMutex
```

Comment:
We can live with a single rwLock for the whole data structure to simplify. It's unlikely to make a noticeable difference to have different locks for each map, especially since the amount of time spent holding the locks is minimal